src/os/bluestore/AvlAllocator.cc

```diff
@@ -0,0 +1,429 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "AvlAllocator.h"
+#include "common/debug.h"
+
+#define dout_context cct
+#define dout_subsys ceph_subsys_bluestore
+#undef  dout_prefix
+#define dout_prefix *_dout << "AvlAllocator "
+
+MEMPOOL_DEFINE_OBJECT_FACTORY(range_seg_t, range_seg_t, bluestore_alloc);
+
+/*
+ * NOTE: caller is responsible for all locking.
+ */
+static int
+range_tree_compare(const void *x1, const void *x2)
+{
+  const range_seg_t *r1 = static_cast<const range_seg_t *>(x1);
+  const range_seg_t *r2 = static_cast<const range_seg_t *>(x2);
+
+  assert(r1->rs_start <= r1->rs_end);
+  assert(r2->rs_start <= r2->rs_end);
+
+  return ((r1->rs_start >= r2->rs_end) - (r1->rs_end <= r2->rs_start));
+}
+
+/*
+ * Comparison function for the private size-ordered tree.
+ * Tree is sorted by size, larger sizes at the end of the tree.
+ */
+static int
+range_tree_size_compare(const void *x1, const void *x2)
+{
+  const range_seg_t *r1 = static_cast<const range_seg_t *>(x1);
+  const range_seg_t *r2 = static_cast<const range_seg_t *>(x2);
+  uint64_t rs_size1 = r1->rs_end - r1->rs_start;
+  uint64_t rs_size2 = r2->rs_end - r2->rs_start;
+
+  int cmp = AVL_CMP(rs_size1, rs_size2);
+  if (likely(cmp))
+    return (cmp);
+
+  return (AVL_CMP(r1->rs_start, r2->rs_start));
+}
+
+range_seg_t *
+AvlAllocator::_find_block(avl_tree_t *t, uint64_t start, uint64_t size)
+{
+  range_seg_t rsearch;
+  avl_index_t where;
+
+  rsearch.rs_start = start;
+  rsearch.rs_end = start + size;
+
+  void *p = avl_find(t, static_cast<void *>(&rsearch), &where);
+  if (p == NULL) {
+    p = avl_nearest(t, where, AVL_AFTER);
+  }
+
+  return static_cast<range_seg_t *>(p);
+}
+
+/*
+ * This is a helper function that can be used by the allocator to find
+ * a suitable block to allocate. This will search the specified AVL
+ * tree looking for a block that matches the specified criteria.
+ */
+uint64_t AvlAllocator::_block_picker(avl_tree_t *t,
+  uint64_t *cursor, uint64_t size, uint64_t align)
+{
+  range_seg_t *rs = _find_block(t, *cursor, size);
+  while (rs != NULL) {
+    uint64_t offset = P2ROUNDUP(rs->rs_start, align);
+
+    if (offset + size <= rs->rs_end) {
+      *cursor = offset + size;
```

Comment:
General concern about such cursor behavior (already highlighted for stupid alocator at #18494) is that it permanently increases that causes excessive fragmentation. Here is a scenario I observed for Stupid Allocator and that's probably is applicable here as well:
There is 40Gb block device (coupled with 16 Gb DB and 10 Gb WAL devices but that's doesn't matter). It's sequentially pre-condifitioned with 32 Gb of data. Hence we have 8 Gb free at block device and allocator has a single continuous 8Gb free extent at this point. Then we start doing random 4K overwrites that cause some 4K extents to return to allocator while new ones are allocated from that large 8Gb extent. Please note that returned small extents aren't reused until we drain all the 8Gb extent since cursor permanently goes up. After overwriting 8Gb of data we have large continuous extent fully exhausted and free 8Gb spread randomly over released 4K extents. There are good chances that existing continuous extents are pretty short (<1Mb). One day bluefs rebalance procedure comes into action and requests that long continuous extent(s) of 1Mb... Allocator fails to do that and we have an assert. That's a real use case I faced recently...
May be we should do a cursor rewind on extent release?